cpp/src/arrow/stl.h

```diff
@@ -35,11 +35,13 @@ namespace stl {
 template <typename T>
 struct ConversionTraits {};
 
-#define ARROW_STL_CONVERSION(c_type, ArrowType_) \
-  template <>                                    \
-  struct ConversionTraits<c_type> {              \
-    using ArrowType = ArrowType_;                \
```

Comment:
This does not remove it, see below. This is just a reindentation of the existing code.

Reply:
The `arrow_type()` method does the same, but at runtime.
The issue was the `ListType<>` template class which produced linker issues on Windows.

What do you need to know at compile-time exactly? Just the type id (e.g. `Type::INT8`)?

Reply:
@pitrou  I would assume yes, compile time resolving of primitive type. Note that I implemented this in a local patch then threw it away because I got around it.

I'd propose that we make this explicit in `type_traits.h` as this was the place I first looked for. Shall I open a jira ticket?

Reply:
Yes, a mapping from a C type to the corresponding arrow DataType class would be useful.

Reply:
Yes, I need to do it at compile compile time, not runtime. Moving it to `type_traits.h` would be ok for me. Notice that I personally do not care about `ListType<>`, all I really need is the code starting from line 38.

Reply:
@ktf see https://issues.apache.org/jira/browse/ARROW-4268